### PR TITLE
Reject attempts at using PCH

### PR DESCRIPTION
**Pull request by kimgr**
IWYU as-is cannot support PCH, because we care deeply about the include graph, and it does not appear to be preserved when Clang interprets the precompiled header format.

We used to fail mysteriously with segmentation faults due to missing files; now we fail with an explicit message on startup.

Remove null-check for this scenario in iwyu:VisitUsingDecl, that code should now be unreachable.

### CODE REVIEW

**Comment by kimgr** [comment]
@kilroyd This is what I had in mind for PCH handling earlier.

@iven @ts826848 A workaround/clarification for #1230 and #1243.

**Comment by kimgr** [comment]
> ### Informational: IWYU dogfood results
> include-what-you-use (exit: 0)
> fix_includes.py (exit: 0)
> diff
> ```diff
> diff --git a/iwyu.cc b/iwyu.cc
> index 1474b78..c32dd33 100644
> --- a/iwyu.cc
> +++ b/iwyu.cc
> @@ -123,7 +123,6 @@
>  #include "clang/Basic/Specifiers.h"
>  #include "clang/Basic/TypeTraits.h"
>  #include "clang/Frontend/CompilerInstance.h"
> -#include "clang/Frontend/CompilerInvocation.h"
>  #include "clang/Frontend/FrontendAction.h"
>  #include "clang/Lex/Preprocessor.h"
>  #include "clang/Sema/Sema.h"
> ```

How cool is this??? :magic_wand:

**Comment by kilroyd** [comment]
IWYU has the option `--pch-in-code`. Does IWYU still support including a PCH through an include directive? Or should that option be removed?

If keeping `--pch-in-code` should we add some documentation about what PCH usage is supported? Also clarify the error message to avoid saying all PCH is not supported.

**Comment by kimgr** [comment]
@kilroyd `--pch_in_code` is a poor way of saying "hey, consider the first `#include` in every .cpp file to be an MSVC-style precompiled header, which can be considered equivalent to a command-line -include".

MSVC used to have this terrible default setup for PCH that the precompiled header was called `"stdafx.h"`, there was a `"stdafx.cpp"`, whose only purpose was to generate the PCH, and then every file explicitly including `"stdafx.h"` would instead get the PCH prepended to the compilation unit.  The nice thing is, if you compile such a .cpp file without special PCH flags, it will just work like a normal include.

(I'm not sure if that's still the case, I haven't used MSVC for many years)

So older MSVC codebases will typically have `#include "stdafx.h"` as the first tokens in every .cpp file, which makes it necessary for IWYU to "see past" it. So `--pch_in_code` essentially makes IWYU treat the first `#include` in every .cpp file as if it was a prefix header (provided on command-line using `-include`), and never removes it. The `--prefix_header_includes` switch can be used to control the policy in the presence of such "God" headers.

I wish there was a better name for that switch, but there it is. The help text tries to explain it tersely.

Anyway, long story short: yes!, some more docs would be nice! :smile:

**Comment by kimgr** [comment]
OK, I started writing up an `IWYUFAQ.md`, but it turned into a bigger thing that I'd rather not publish as part of this PR. I'll follow up with a separate thing explaining why it doesn't work, possible ways to tweak the invocation to make it work, and how `--pch_in_code` relates to all this.

**Comment by kilroyd** [comment]
No problem, sounds good to me.

The commit itself seems fine.

**Comment by firewave** [comment]
> > ### Informational: IWYU dogfood results
> > include-what-you-use (exit: 0)
> > fix_includes.py (exit: 0)
> > diff
> 
> How cool is this??? 🪄

🥳

As the exitcodes are `0` if there is a diff it would be helpful to add an indicator to the title so you do not have to check by hand if there is an actual diff.

**Comment by kimgr** [comment]
@firewave Yeah, maybe that would be nice. I think maybe I've spent enough time on this stuff now, though. Should be a small patch to `iwyu-dogfood.bash` if you're feeling up for it :)
